python/ray/train/tests/test_gpu.py

```diff
@@ -108,23 +115,30 @@ def train_fn():
         )
 
 
-# TODO: Refactor as a backend test.
 @pytest.mark.parametrize("num_gpus_per_worker", [0.5, 1, 2])
 def test_torch_get_device_dist(ray_2_node_2_gpu, num_gpus_per_worker):
     @patch("torch.cuda.is_available", lambda: True)
     def train_fn():
-        return train.torch.get_device().index
-
-    trainer = Trainer(
-        TorchConfig(backend="gloo"),  # use gloo instead of nccl,
-        # since nccl is not supported on this virtual gpu ray environment
-        num_workers=int(4 / num_gpus_per_worker),
-        use_gpu=True,
-        resources_per_worker={"GPU": num_gpus_per_worker},
+        session.report(dict(devices=train.torch.get_device().index))
+
+    class TorchTrainerPatched(TorchTrainer):
+        def _report(self, training_iterator) -> None:
+            for results in training_iterator:
+                tune.report(results=results)
```

Comment:
let's move this to the top of the file and also add a description on why we need it?